### ISSUE TITLE
Compatible with ONT data

### ISSUE THREAD

**Issue by OxPen** [Oshlack/JAFFA#39]
Hello,

Just to confirm, is JAFFA suitable for fusion detection with Oxford Nanopore cDNA seq data?

Thanks!!

Peny

**Comment by nadiadavidson** [comment]
Hi Penny,

I've never tried, but in theory if you can get the reads into a fasta format it should work. The error rate might have some impact on the sensitivity of fusion calls, but I think it's worth a try. We would be interested to find out how you go if you do try to run JAFFA on Oxford Nanopore, as I'm sure others would be interested in trying it too.

Cheers,
Nadia.

**Comment by OxPen** [comment]
Thank you very much Nadia, I will update you as soon as I make it work! 

I do have this problem though, duting the Stage_fet_final_list:

```
[1] "Getting the location of fusion transcripts in the genome.."
Calculating gap size in the genome...
Checking if the fusions are in frame...
Merging with read coverage data...
Error in `$<-.data.frame`(`*tmp*`, "known", value = "-") : 
  replacement has 1 row, data has 0
Calls: $<- -> $<-.data.frame
Execution halted
ERROR: Command failed with exit status = 1 : 

R --vanilla --args 1.fastq/1.fastq_genome.psl 1.fastq/1.fastq.reads /home/hassan/panagiota/JAFFA/hg38_genCode22.tab /home/hassan/panagiota/JAFFA/known_fusions.txt 10000 NoSupport,PotentialRegularTranscript 1.fastq/1.fastq.summary < /home/hassan/panagiota/JAFFA/make_final_table.R 


========================================= Pipeline Failed ==========================================

One or more parallel stages aborted. The following messages were reported: 

Branch 1.fastq in stage Unknown reported message:

Command failed with exit status = 1 : 

R --vanilla --args 1.fastq/1.fastq_genome.psl 1.fastq/1.fastq.reads /home/hassan/panagiota/JAFFA/hg38_genCode22.tab /home/hassan/panagiota/JAFFA/known_fusions.txt 10000 NoSupport,PotentialRegularTranscript 1.fastq/1.fastq.summary < /home/hassan/panagiota/JAFFA/make_final_table.R

Use 'bpipe errors' to see output from failed commands.
```

...although I downloaded and cat-ed the correct genome, according to your instructions. Do you have any ideas on how I can get around it please?

Many thanks,

Peny

**Comment by nadiadavidson** [comment]
Sorry, I know it's been a long time to reply to this! I've been on maternity leave.

It sounds like JAFFA has failed to pick up any fusions (which might be because it's not designed for the errors in nanopore data). We would love for it to be able to run on data like this. Do you know of any public or other datasets we could use for testing? I'd love to try it out.

Cheers,
Nadia.

**Comment by OxPen** [comment]
Hi Nadia,

Sorry it took me so long to respond too. I am sorry but I don't know of any repository with ONT data to use for testing. :/ 

Just to let you know, I've run again the program with my sample data and the demo data that you provide and I am stuck in the same step in both:

Stage filter_transcripts

real	0m0.024s
user	0m0.002s
sys	0m0.000s
ERROR: Command failed with exit status = 127 : 

time R --vanilla --args BT474-demo_2.fastq/BT474-demo_2.fastq.psl BT474-demo_2.fastq/BT474-demo_2.fastq.txt 1000 /home/hassan/panagiota/JAFFA/hg38_genCode22.tab < /home/hassan/panagiota/JAFFA/process_transcriptome_blat_table.R &> /home/hassan/panagiota/JAFFA/tools/bin/BT474-demo_2.fastq/log_filter 

Can you help me deal with this please? 
Many thanks,
Peny

**Comment by nadiadavidson** [comment]
Hi Peny,

I've actually found that SRA has some ONT data to play with (although haven't had time to run JAFFA on it yet).

Can you tell me what output is in the file /home/hassan/panagiota/JAFFA/tools/bin/BT474-demo_2.fastq/log_filter? This may help me to solve the issue.

Cheers,
Nadia

**Comment by OxPen** [comment]
Hi Nadia,

Managed to resolve the error at that stage after looking at the log_filter file myself- it was the IRanges library, which for some reason has not been installed.

I ran the pipeline again and now it gets stuck at the get_final_list step, where I get this error:

Error in read.table(file = file, header = header, sep = sep, quote = quote,  : 
  no lines available in input
Calls: read.delim -> read.table
Execution halted
ERROR: Command failed with exit status = 1 : 

R --vanilla --args BT474-demo_2.fastq/BT474-demo_2.fastq_genome.psl BT474-demo_2.fastq/BT474-demo_2.fastq.reads /home/hassan/panagiota/JAFFA/hg38_genCode22.tab /home/hassan/panagiota/JAFFA/known_fusions.txt 10000 NoSupport,PotentialRegularTranscript BT474-demo_2.fastq/BT474-demo_2.fastq.summary < /home/hassan/panagiota/JAFFA/make_final_table.R 

Error in read.table(file = file, header = header, sep = sep, quote = quote,  : 
  no lines available in input
Calls: read.delim -> read.table
Execution halted
ERROR: Command failed with exit status = 1 : 

R --vanilla --args BT474-demo_1.fastq/BT474-demo_1.fastq_genome.psl BT474-demo_1.fastq/BT474-demo_1.fastq.reads /home/hassan/panagiota/JAFFA/hg38_genCode22.tab /home/hassan/panagiota/JAFFA/known_fusions.txt 10000 NoSupport,PotentialRegularTranscript BT474-demo_1.fastq/BT474-demo_1.fastq.summary < /home/hassan/panagiota/JAFFA/make_final_table.R 

I see in my directory that the size of the .psl file is 0, I assume that this has to do with it.

Sorry for the trouble :/

Many thanks

Peny

**Comment by nadiadavidson** [comment]
Hi Peny,

Yes a .psl file size of 0 is definitely bad! Are they the BT474-demo_1.fastq_genome.psl files or the BT474-demo_1.psl files with 0 size? If it's the _genome ones, it could be something to do with the reference genome being wrong. Just to rule it out, did you download the hg38 reference from UCSC?

I also just realised that you're running the BT474-demo_1 and BT474-demo_2 files separately, which is completely fine, but would mean that you either needed to set readLayout=single or convert the files to .fasta format. Just thought I would check this.

For the nanopore data, it looks like you might need to play a little bit with the blat settings to help the alignment work with the high error rate. e.g. could try something like:

bpipe run -p tileSize=11 -p minScore=100 -p minIdTrans=85 -p readLayout=single JAFFA_direct.groovy <data>
see https://genome.ucsc.edu/goldenpath/help/blatSpec.html for what these three parameters mean. minIdTrans is just minIdentity.

Cheers,
Nadia.

**Comment by OxPen** [comment]
Hi Nadia,

It's the genome psl file which is empty but I am 100% sure that I've used the correct genome ref from UCSC, as I followed the directions step by step. 
I have set the readLayout to single in the stages file so I think this shouldn't be the problem.

I have downloades again the ref genome and hg38_genome_annotation and I am running it again but it's

**Comment by OxPen** [comment]
Hi Nadia,

It is indeed the genome.psl file which is empty. I have re-installed the ref file, following the directions very carefully but still I get the same, an empty psl file. Also, I have to manually enter many of the commands, when errors pop out and then rerun the command from the beginning, in order to proceed.

Any ideas how to deal with this? :(

Best

Peny

**Comment by nadiadavidson** [comment]
Sorry to hear that you're having trouble still. Can you post a list of the file sizes ("ls -l BT474-demo_1.fastq/") and I will check against my demo files to see if something has gone wrong upstream of the .psl file.

**Comment by OxPen** [comment]
Hi Nadia,

Yes, the demo files sizes are:

-rwxr-x--- 1 path1030 hassan group 6745785 Apr 10  2014 BT474-demo_1.fastq.gz
-rwxr-x--- 1 path1030 hassan group 6869178 Apr 10  2014 BT474-demo_2.fastq.gz
-rwxr-x--- 1 path1030 hassan group 6665119 Apr 10  2014 MCF7-demo_1.fastq.gz
-rwxr-x--- 1 path1030 hassan group 6764825 Apr 10  2014 MCF7-demo_2.fastq.gz

**Comment by nadiadavidson** [comment]
Ah sorry, I meant the sizes of the files created by JAFFA. These should be in the directory BT474-demo_1.fastq/ etc.

**Comment by OxPen** [comment]
Hi Nadia, happy New Year!

Sorry it took me so long to get back to you but I just managed to make Jaffa work for the demo files (I had to re-install it and do everything from the begining).
As I said, it now runs with the demo but when I try to run it with my ONT data I get this:

```
================================== Stage get_final_list (merged) ===================================
/software/TOOLS/R/R-3.5.1/bin/exec/R: Symbol `rl_readline_state' has different size in shared object, consider re-linking
R version 3.5.1 (2018-07-02) -- "Feather Spray"
Copyright (C) 2018 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.
  Natural language support but running in an English locale
R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.
Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.
> options(echo=FALSE)
[1] "Getting the location of fusion transcripts in the genome.."
Calculating gap size in the genome...
Checking if the fusions are in frame...
Merging with read coverage data...
Error in `$<-.data.frame`(`*tmp*`, known, value = "-") : 
  replacement has 1 row, data has 0
Calls: $<- -> $<-.data.frame
Execution halted
ERROR: Command failed with exit status = 1 : 
```

I have downloaded all the refs properly, as advised. Do you have any idea why this is happening?

Many thanks,

Peny

**Comment by nadiadavidson** [comment]
Hi Peny,

This error looks like JAFFA has not found any fusions (sorry it should exit nicely and give a messsage, but doesn't do that currently).

Whether JAFFA can find anything with ONT data is still experimental, the main issue being the error rate of course. You might have some luck if you modify the settings for alignment. e.g. running the pipeline like this:
<bpipe_dir>/bpipe run -p tileSize=11 -p minScore=50 -p minIdTrans=85 -p readLayout=single <JAFFA_dir>/JAFFA_direct.groovy <data>
This sets the blat parameters tileSize, minScore and minIdentity(%)  respectively.

I'm looking at this with some public ONT data to see how JAFFA performs.
Cheers,
Nadia.

**Comment by OxPen** [comment]
Hi Nadia,

Thank you for your response- I just tried it again with no luck.
My guess is that it's due to bowtie2, perhaps using another aligner like minimap2 or NGMLR would make more sense with ONT data.

Please let me know if you think of anything-it would be the only gene fusion tool that actually works with ONT!

Thanks again,

Peny

**Comment by OxPen** [comment]
Dear Nadia,

Thank you very much for this! Can I please ask you to point me to the data you used so I can give it a try with those too?

I will let you know how it goes with my own too!

Thanks!!
Peny

**Comment by nadiadavidson** [comment]
Hi Peny,

The dataset I used was from this paper, https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5726485/. The sample with SRA ID DRR059313 is a mixture of cell lines and has 7 known fusions. JAFFA picks up 6 of them with the setttings provided above. It actually picks up a few different splice forms of the fusions too which is cool. 

The fusion that's missed has only low expression and probably the blat settings need to be tweaked some more, but I'm very happy with the performance as a first attempt. Interestingly, the missed fusion is picked up in the first stages of fusion detection and then later filtered out as a false positive. The file <sample>.txt lists all these preliminary fusions. If you run something like:
cut -f4  DRR059313.fastq/DRR059313.fastq.txt | sort | uniq -c
You can see what was called. The number indicates read support. The fusions are not ordered correctly at this stage, so you will often see GeneA:GeneB and GeneB:GeneA, which are likely to be the same fusion. This list also doesn't give any other information like break point position. But might be useful to see if a fusion you're expecting has been filtered out later.

The final list of fusions which survive all filtering are given in <sample>/<sample>.summary as well as jaffa_results.csv. I'm running on another dataset now and will see if I pick anything up. Would love to know how you go. I may write this up on our wiki at some stage in case others would like to use JAFFA on this sort of data.

Cheers,
Nadia.

**Comment by OxPen** [comment]
Hi Nadia,

Thanks, that was very helpful!! I run it too on the public dataset and indeed I got the fusions!
However, in two out of three samples of mine I got an error on the final stage:

options(echo=F)
Compiling the results from:
all
Done writing output jaffa_results.csv
Done writing jaffa_results.fasta
All Done
ERROR: Expected output file jaffa_results.fasta could not be found 

I also checked the .summary file but it was empty, apart from the headers. Does that mean it didn't pick up anything? In the sample.txt file where the prelim fusions are saved, I can't find the one fusion I am expecting to find, which should be EWS-FLI1.

Many thanks,

Peny
